### ISSUE TITLE
BaseOptionModel should serialize to the closest data type by default instead of binary

### ISSUE THREAD

**Issue by reduckted** [VsixCommunity/Community.VisualStudio.Toolkit#112]
When saving a property that is a `bool`, the `BaseOptionModel` should store the value as a boolean using `WritableSettingsStore.SetBool()` instead of using a binary serializer and storing it as a string. 

The same should occur to `int`, `uint`, `long`, `ulong` and `string`.

**Comment by madskristensen** [comment]
Why is this better than using serializing?

**Comment by StevenBonePgh** [comment]
For one it is more performant to use the native type if possible. The second reason is that migrating an extension where the `WritableSettingsStore` was used in this manner to use `BaseOptionModel` instead, previously stored values of the specific type will fail to be read or deserialized. I added the `OverrideDataTypeAttribute` in #29 that can be applied to the property to specify the storage type, which may be a workaround. 

This suggestion, as proposed, would be a breaking change to anyone already using `BaseOptionModel` as-is that does not specify this attribute. Some built-in types, like byte and enumerations are edge cases for determining the closest native type. Then there is float, double, datetime, and custom types where current serialization would likely be the best. Could be made non-breaking if `BaseOptionModel ` is used as a base class and a new `BaseOptionModelEx` class could override the storage methodology - this would also allow for the `BaseOptionPage` to work with either implementation.

**Comment by madskristensen** [comment]
I think we should only use a single method for storing the values. If that means using the right/closes data types, then let's get rid of the serialization completely. Would all other data types such as `DateTime` not just be stored as strings in this case?

**Comment by madskristensen** [comment]
I'm not too concerned about breaking existing extensions on this one since the toolkit is still a -pre package. I'd rather we focus on getting it right. 

Will this change also mean that the attributes you added a while back could be removed?

**Comment by StevenBonePgh** [comment]
I don't know of a reasonable way, other than the default serializer you already use, to handle custom or built in non-natively supported types. For commonly used types that are not exact matches we can hard-code how these are handled. I think for byte/short/ushort or enums with those underlying types we can promote to store as integer, char as string (or possibly short), datetime as long/Int64, and standard supported types as themselves, and for anything else we could use your serializer.

I'd suggest keeping the override attribute for porting from existing extensions where the type we use/infer by default is not the proper one, and for extensions like Tweakster that already use this you can apply the attribute to use the serializer to your existing properties.

I can give this one a go, perhaps a PR for it by Monday?

**Comment by madskristensen** [comment]
Sounds good. I'm ok breaking Tweakster settings with this change. I'd rather have a simple, easy to reason about, one-way of doing things, API in the toolkit.

**Comment by StevenBonePgh** [comment]
I've made these changes locally and have a few concerns to discuss. It was relatively simple to support all of the [integral numeric types](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types) and: float, double, decimal, char, bool, and enums of any underlying type. If you noticed, the commonality behind these types is they are all value types.

- DateTime: Current thoughts are to store as Int64 using `ToBinary/FromBinary` which round-trips perfectly, including Kind. I was thinking to use Convert.ChangeType and store as string, although this does not round-trip with 100% accuracy as it will not maintain 'Kind' and is off by a couple of milliseconds. Any objections to the Int64 storage of DateTime?
- String: as this is a reference type, it can be null. You can't store null in the SettingsStore. You also can't infer, by absence of the setting (SettingsStore.PropertyExists) that this _means_ null, as it is possible that the setting was never saved and therefore the property should not be set on the object during load. A few options: 
--delete setting if null thereby a subsequent load will use the object's current value rather than the last non-null value (optional to log an error in this case). Essentially, null would not be supported.
--use a magic value like a string containing a null character to indicate null if this is possible (have not explored this, but would make a null string a supported possibility though it means a string containing a null character will result in this behavior).
- Other types: Use serialization method from BaseOptionModel. In here, for reference types, we can infer that an 'empty' serialized value means null. See below.

For unknown types, the current serialization mechanism in BaseOptionModel is the `BinaryFormatter`. This has fallen out of favor due to security concerns. For this I have a few suggestions:
Signature Change: rather than using a string to store serialized data, I think the binary MemoryStream storage is the way to go. Also, the type and property name should be included for both serialization and deserialization methods allowing for these to be overridden with logic specific for type and/or specific property names. They would look like this:
```
protected internal virtual object? DeserializeValue(MemoryStream memoryStream, Type type, string propertyName)
protected internal virtual MemoryStream SerializeValue(object? value, Type type, string propertyName)
```
Serializer: We are limited by target framework (4.6 and 4.72), and the desire to avoid or minimize the use of external libraries for multiple reasons.   If we wish to avoid `BinaryFormatter` as a default choice, given the security implications, that pretty much leaves `XmlSerializer` unless I am missing something. Thoughts?

**Comment by madskristensen** [comment]
DateTime as Int64 I'm cool with

Can we make strings empty instead of `null` upon saving/loading from the settings store?

I'm cool with using the XML serializer. But it did get me thinking, do we need support for complex types instead of just the natively supported types by the settings store? I don't think it's common to use other types, but if it is then the extender should handle serialization manually and store as string. Thoughts?

Unit test project, yes please!!!! Thank you. I can hook up the CI part

**Comment by StevenRasmussen** [comment]
> Finally, any objections to including a Community.VisualStudio.Toolkit.UnitTests project? I started one using MsTest as the SettingsStore is abstract it is easy to use Moq for testing, and to test the core part of this change does not require any bits from an active instance of visual studio. I lack the know-how for integrating this into CI, but I am sure there are others here who could help with that.

> Unit test project, yes please!!!! Thank you. I can hook up the CI part

I guess my opinion would be to use `XUnit` for unit testing as opposed to `MSTest`.  This *seems* to be the direction that MS is going internally as both the Asp.Net Core and EFCore projects use it.  I've also found [Fluent Assertions](https://github.com/fluentassertions/fluentassertions) to be extremely useful.  Just my 2 cents 😉

**Comment by yannduran** [comment]
+1 for both xUnit and FluentAssertions

**Comment by yannduran** [comment]
@madskristensen I think the most _exotic_ things that I've seen done in settings would be lists of strings and lists of colors. Both of those should be a problem, right?

**Comment by madskristensen** [comment]
A list of strings should potentially just be joined into a single string with a separator. Is there UI support for string arrays in a `DialogPage`?

**Comment by madskristensen** [comment]
To sum up my thinking on this:

* Only support data types that have simple type representations such as `string` and `decimal`. This includes enums which are `int`
* Throw exception when other types are used. The extender should convert to the right type in their own code
* Don't provide the overriding attributes. There should be just one way settings are stored
* Back-compat is not a goal. The Toolkit's focus is on new extensions and making it as simple as possible for extenders.

**Comment by StevenBonePgh** [comment]
Just raised a PR on this. Thanks to all for your feedback. Details in the PR, but I'll mention a few things here:

- @StevenRasmussen @yannduran Thanks for the suggestion, converted what I had to xUnit, was already using FiuentAssertions so it was super easy.
- @yannduran I can't believe I initially missed Color! Need that for own extension. Using To[From]Argb to store as int. Lists, however, will get the XmlSerializer treatment.
- @madskristensen null strings go in as empty as suggested. Unit test project is in place. And string[] is supported by DialogPage - I added it to the sample, and is using XmlSerializer.

Other thoughts: I agree with one-way and simple. As this is a developer toolkit, I also don't want to be handcuffed and realize because I bought-in to a toolkit to make thing easier it also has me forced into to one way of thinking and is not itself extensible. As the code was already complete sans tests and comments, I left the overriding attributes in - there is no cost to having the capability. I also personally need some of them. The new DI framework and conversation about interfaces, including the need for them when unit testing, further illustrates the point that extensibility is desired.
